File: lab15_3/CycleList.cs
using System;
using System.CodeDom;
using System.ComponentModel.Design;
using System.Runtime.InteropServices;
using System.Runtime.Remoting.Messaging;

namespace lab15_3
{
    public class Node<T>
    {
        public T data;
        public Node<T> next;
        public Node<T> perv;
    }
    
    public class CycleList<T>
    {
        public Node<T> head;
        public CycleList()
        {
            this.head = null;
        }

        public void Append(T value)
        {
            if (this.head == null)
            {
                this.head = new Node<T> {data = value};
                return;
            }

            if (this.head.next == null)
            {
                this.head.next = new Node<T> {data = value};
                this.head.perv = this.head.next;
                this.head.next.next = this.head;
                this.head.next.perv = this.head;
                return;
            }

            var tmp = this.head;
            while (tmp.next != this.head)
                tmp = tmp.next;
            tmp.next = new Node<T> {data = value};
            tmp.next.perv = tmp;
            tmp.next.next = this.head;
            this.head.perv = tmp.next;
        }

        public int Count()
        {
            var tmp = this.head;
            int count = 0;
            while (tmp.next != head)
            {
                tmp = tmp.next;
                count++;
            }

            return count + 1;
        }

        public T Pop(int index)
        {
            T res;
            
            if (index == 0)
            {
                res = this.head.data;
                var last = this.head.perv;
                this.head = this.head.next;
                this.head.perv = last;
                return res;
            }

            if (index >= this.Count())
                throw new Exception("Incorect index");
            int count = 0;
            var tmp = this.head;
            while (count != index)
            {
                tmp = tmp.next;
                count++;
            }

            res = tmp.data;
            var newPrev = tmp.perv;
            var newNext = tmp.next;
            newPrev.next = newNext;
            newNext.perv = newPrev;
            return res;
        }
    }
}

File: python/task3/cycleList.py
import random


class Node:

    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class CycleList:

    def __init__(self):
        self.head = None

    def append(self, value):
        if self.head == None:
            self.head = Node(value)
            return
        if self.head.next == None:
            self.head.next = Node(value)
            self.head.prev = self.head.next
            self.head.next.next = self.head
            self.head.next.prev = self.head
            return
        tmp = self.head
        while tmp.next != self.head:
            tmp = tmp.next
        tmp.next = Node(value)
        tmp.next.prev = tmp
        tmp.next.next = self.head
        self.head.prev = tmp.next

    def count(self):
        tmp = self.head
        count = 0
        while tmp.next != self.head:
            tmp = tmp.next
            count += 1
        return count + 1

    def pop(self, index):
        if index == 0:
            res = self.head.data
            last = self.head.prev
            self.head = self.head.next
            self.head.prev = last
            return res

        if index >= self.count():
            raise Exception("Incorrect index")
        count = 0
        tmp = self.head
        while count != index:
            tmp = tmp.next
            count += 1
            # print(tmp)
        result = tmp.data
        newPrev = tmp.prev
        newNext = tmp.next
        newPrev.next = newNext
        newNext.prev = newPrev
        return result
if __name__ == "__main__":
    Surnames = {"Мельник", "Шевченко", "Коваленко", "Бондаренко", "Бойко", "Ткаченко", "Кравченко", "Ковальчук", "Коваль",
                "Олійник", "Шевчук", "Поліщук", "Ткачук", "Савченко", "Бондар", "Марченко", "Руденко", "Мороз", "Лисенко",
                "Петренко", "Клименко", "Павленко", "Кравчук", "Кузьменко", "Шило"}
    cycled1 = CycleList()
    cycled2 = CycleList()
    for i in range(10):
        cycled1.append(Surnames.pop())
        cycled2.append(Surnames.pop())
    for i in range(10):
        print(f"{cycled1.pop(random.randint(0,cycled1.count()-1))} VS {cycled2.pop(random.randint(0,cycled2.count()-1))}")

